### ISSUE TITLE
Foot routing should give higher preferences for main streets with sidewalks

### ISSUE THREAD

**Issue by elibar** [graphhopper/graphhopper#633]
In the foot routing there is preferences to side roads over main roads for safety reasons.
However if  a sidewalk exists the preferences should be equals or even higher then a side road.

For example, in the link below, Oxford street, like many other main streets, is of 'Primary Road' type and has `sidewalk=both` or `sidewalk=left|right` tags. The foot routing prefers longer route with turns through "Alley" and "Territory Road"   over a shorter path with no turns in 'Primary Road'.

Example link: https://graphhopper.com/maps/?point=51.514885%2C-0.143766&point=51.517236%2C-0.125055&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

**Comment by ratrun** [comment]
This behaviour is intended. In case you prefer to walk along 'highway=primary' you can create a branch.

**Comment by karussell** [comment]
I think preferring 'highway=primary' is not what was requested. The problem here is more that although it is safe due to sidewalk tags still the primary roads are avoided. And the question is: is this the behaviour how it should be?

**Comment by elibar** [comment]
I do not wish to walk on highways, the thing is that many streets in cities that are perfectly good and safe for walking are marked as primary. See the example in the description, I checked it also on other cities and the main streets where there is shops, restaurants and good light are marked as primary or secondary.  
Having sidewalks is a good indication that the street is safe to walk and shouldn't be avoid.

Here some more examples

London
https://graphhopper.com/maps/?point=51.514752%2C-0.14441&point=51.51769%2C-0.119519&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale
https://graphhopper.com/maps/?point=51.575229%2C-0.196424&point=51.580617%2C-0.198526&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale
https://graphhopper.com/maps/?point=51.514752%2C-0.14441&point=51.516448%2C-0.130377&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

Manhattan
https://graphhopper.com/maps/?point=40.775943%2C-73.947151&point=40.778689%2C-73.953727&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

Berlin
https://graphhopper.com/maps/?point=52.517331%2C13.390789&point=52.516913%2C13.383622&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale
https://graphhopper.com/maps/?point=52.517853%2C13.388686&point=52.511324%2C13.389738&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

Tel Aviv
https://graphhopper.com/maps/?point=32.073929%2C34.7663&point=32.078284%2C34.767619&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

**Comment by ratrun** [comment]
Even if a primary has a sidwalk, I can't believe that may people prefer walking there in case that there are nicer, but longer alternatives. We already had a pretty long #476 to reach this goal.
I haven't tested it, but eventually a dynamic prioBoost level based on PR #630 with a small value would return your prefered result.

**Comment by elibar** [comment]
hi ratrun, 
Take a look at the routes in the example, they are not nicer then the obvious “continue straight” route, they are longer and harder to follow. I tested hundreds of pedestrian routes and I got many “funny” routes. In many cases I got routs where I wouldn’t want to go there alone at night, while at the same time there are other shorter, better routes with good sidewalk. 

We can debate for days if it is nicer to go through the alleys and enjoy the secretes of the city or to go on the main streets and enjoy the showcases, but we deals with “weighting=fastest” so I think that we can agree that for these weighting we look for the fastest route, as long as we do not put the pedestrians in danger of walking in places where they shouldn’t walk in.

Compare the GH implementation to the ones of Google and Here, they do not suggest routes, certainly not as their first suggestion, that walk extra distance just to avoid main streets where sidewalks exists. I do not say that just because Google and Here implemented it that way GH should implement it the same, I just say that maybe they have a point.

I also had the opportunity to let real pedestrians walk according to the routes produced by the current version and I got similar feedback from significant percentages of them.

To summarize, what I’m saying is that it is clear that the current routes are not optimized in many cases. It might be that the solution is different than the one I wrote in the title, but I think we should discuss how to optimize it, rather than if there is a need to optimized it.

**Comment by ratrun** [comment]
I have tested the first Berlin case https://graphhopper.com/maps/?point=52.517331%2C13.390789&point=52.516913%2C13.383622&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale with my current master from https://github.com/ratrun/graphhopper/tree/master, which enables to dymanically change the priority boosting level with a slider. So here you can tune between nice and fast. niceLevel=0.28 gives the result you prefer. This installation requires the usage of the flexibility mode.
As expected, using the same niceLevel level with the Aberdeen issue from  #476 breaks the result there.

Please note that even if the URL parameter contains 'weighting=fastest`, internally the priority weighting is used in case that an encoder supports it. As the foot encoder supports it, the priority weighting is always used here.

I believe that the flexibility mode should include support for dynamic changing of the prio boosting level, similar to my current master. I intend to make a further PR for this in case that #630 gets merged.

But this does not work for the CH mode. As far as I understood Peter, the announced improvements for the flexibility mode will enable the pre-calculation of graphs with different niceLevel values. This would make switching between two modes possible in CH installations. Only Peter can give an answer if he has the server ressources to host such two CH foot profiles. Please note that my main interest is on the flexiblity mode as I'm running graphhopper on a local installation.

**Comment by karussell** [comment]
What do you think if we move the current `foot` to a new profile `hike` and remove certain priorizations for `foot`, like preferring the (hiking) networks and lowering some other 'beauty road' aspects. So that we have at the end two profiles
 * 'foot' (fastest but still safe) and
 * 'hike' with all the priorizations, tourism preferences and also potentially more allowed paths (allow higher grades?)

**Comment by elibar** [comment]
I like it, it seems that indeed we're dealing with two walking profiles

**Comment by ratrun** [comment]
Of course such a change would be an easy solution. Eventually in this case the URL parameter `vehicle` needs to  get renamed into 'profile'. This would be a big change, as it breaks the external API. 
Anyhow, I suggest to wait with such a change for the promised improved flexibility mode, as at least the first step for that change should be a simple and easy to understand example use-case for this extension.
I believe that there are more differences between a real hiking profile and a nice foot vehicle. One example is the weighting of sac_scale and surface, which are probably not the same for a nice foot and a real hiking profile.

**Comment by karussell** [comment]
> Eventually in this case the URL parameter vehicle 

The parameter vehicle would not change

> I believe that there are more differences between a real hiking profile and a nice foot vehicle. 

Sure, we could improve them later on. I would propose to make the Foot encoder the super class and let the hiking profile extend it making it possible to overwrite or change certain properties.

Do you think that this change could do any harm?

**Comment by neiljp** [comment]
As the original reporter of the "Aberdeen issue", I was going to report another, only to find this one. The case I have is a simple one of:
https://graphhopper.com/maps/?point=57North%20Hacklab%2C%20AB11%205BN%2C%20Aberdeen%2C%20United%20Kingdom&point=57.146391%2C-2.099344&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale
All of Union Street has sidewalk on both sides - it is essentially a/the main shopping street, but with the road itself being a major route (UK A-road). I don't see a way that it should be tagged differently in a better way - the sidewalks are certainly not sufficiently separated to warrant separate tagging. Even so, the above route... well it's just not one that I'd ever want out of an engine, except maybe one that said 'avoid pollution' (ie. traffic) or 'via pubs'!

**Comment by elibar** [comment]
Totally agree with neiljp, the current weights makes the foot routes not usable, certainly not in populate areas.

**Comment by karussell** [comment]
Please have a look at the new `foot` profile. (And the new `hike` is the old `foot`)

The new `foot` profile is certainly not perfect but does not avoid primary roads anymore if they have a sidewalk, also other preferences are lowered a bit. Please give feedback about this change!

Also now that you can disable CH on demand you can try `ch.disable=true&weighting=shortest` and see if this satisfies your needs.

**Comment by karussell** [comment]
Also I chose 'hike' ... does anyone prefer 'hiking' or something else?

**Comment by neiljp** [comment]
Name: Don't most of the other profiles have the simple form, ie. car, not 'driving', bike, not 'biking' or 'bicycling' ? Admittedly foot or hike are not 'vehicles' in quite the same way ;)

Testing: It sounds good, potentially, though I'm not in a position to test this for a bit - maybe once I've moved house and am settled! It would be interesting to see if it makes the route I linked above more 'obvious', and possibly also the previous example around the edge of the roundabout, where it still took a detour via steps and footpaths away from the roundabout, rather than following the parts of (major road) roundabout with a good sidewalk (after the previous changes).

**Comment by karussell** [comment]
> Name: Don't most of the other profiles have the simple form, ie. car, not 'driving', bike, 
> not 'biking' or 'bicycling' ? Admittedly foot or hike are not 'vehicles' in quite the same way ;)

Thanks - I'll take this as a vote for `hike` :) 
and yes: we'll have to rename `vehicle` to `profile` at some point for a more consistent naming :)

> and possibly also the previous example around the edge of the roundabout,

what roundabout? [this here](https://graphhopper.com/maps/?point=57.149389%2C-2.090321&point=57.149691%2C-2.092541&vehicle=foot)

**Comment by neiljp** [comment]
Preview-link just wiped out my reply, but:
- no idea of internals, but name consistency is good :)
- roundabout side 1:
https://graphhopper.com/maps/?point=57.154006%2C-2.103474&point=57.153413%2C-2.100132&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=OpenStreetMap
- roundabout side 2:
https://graphhopper.com/maps/?point=57.154949%2C-2.10187&point=57.154393%2C-2.099456&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=OpenStreetMap

Both of the roundabout examples there would be better (more direct, and actually flatter terrain) just following the edge of the roundabout, rather than along the cycleways/steps.

**Comment by karussell** [comment]
> Both of the roundabout examples there would be better 

Yes, especially as those primary roads seem to contain acceptable sideways. The new `hike` profile will fix the second example and [do this](https://graphhopper.com/maps/?point=57.154949%2C-2.10187&point=57.154472%2C-2.100422&point=57.154393%2C-2.099456&vehicle=foot). The first example seems to me acceptable even for the foot profile (in both profiles we currently do not take elevation into account, in `hike` we could think about it)

Will merge this issue soonish before the [first release candidate of 0.7](

Both of the roundabout examples there would be better )

**Comment by karussell** [comment]
This is now deployed to the API & GraphHopper Maps. See [foot](https://graphhopper.com/maps/?point=57.154949%2C-2.10187&point=57.154393%2C-2.099456&vehicle=foot) vs. [hike](https://graphhopper.com/maps/?point=57.154949%2C-2.10187&point=57.154393%2C-2.099456&vehicle=hike)

**Comment by neiljp** [comment]
This roundabout is a great test case, apparently ;)
The links in the previous comment are a great improvement and show the route I would take over that short distance.
However (ahem!) I tried dragging the red marker to further around the roundabout, onto Gallowgate, and although it takes the good shortcut still (not the steps), it also takes a section which has sidewalk=no, which is avoided in the routes in the above comment, as follows:
https://graphhopper.com/maps/?point=57.154949%2C-2.10187&point=57.153171%2C-2.099757&locale=en-US&vehicle=foot&weighting=fastest&elevation=true&layer=Omniscale

**Comment by karussell** [comment]
Will look into it and you mean it should avoid [this way](http://www.openstreetmap.org/way/272280569) (?)

**Comment by neiljp** [comment]
Hi @karussell; yes, I actually meant to link to that way!

**Comment by karussell** [comment]
@neiljp I've created a new issue: #744
